onnxruntime/contrib_ops/cpu/transformers/beam_search_impl_gpt.h

```diff
@@ -202,15 +228,51 @@ Status BeamSearchGpt<T>::Execute(const FeedsFetchesManager* init_run_feeds_fetch
                   parameters->vocab_size,
                   parameters->sequence_length,
                   parameters->max_length,
+                  gpt_subgraph_.num_layers,
+                  gpt_subgraph_.num_heads,
+                  gpt_subgraph_.head_size,
                   parameters->output_scores,
-                  use_position);
+                  use_position,
+                  use_preallocated_past_and_present_buffers_);
 
   init_beam_state_func_(&beam_state,
                         cpu_state.sequence_lengths,
                         parameters->batch_size,
                         parameters->num_beams,
                         this->ort_stream_);
 
+  // Allocate present state fetches.
+  // We will use the same buffer always - it is pre-allocated to hold max_sequence_length of present state values.
+  if (use_preallocated_past_and_present_buffers_) {
+    // init run and other run subgraphs are expected to have same past/present info. So, use any one of them below.
+    fetches.reserve(static_cast<int64_t>(gpt_subgraph_.GetFirstPresentOutputIndex()) + gpt_subgraph_.num_layers);
+    fetches.resize(gpt_subgraph_.GetFirstPresentOutputIndex(), OrtValue());
+    for (int64_t layer = 0; layer < static_cast<int64_t>(gpt_subgraph_.num_layers); layer++) {
+      OrtValue present_tensor;
+      gsl::span<T> present_state_buffer = beam_state.GetPresentStateBuffer();
+
+      gsl::span<T> current_layer_present_buffer = present_state_buffer.subspan(static_cast<int>(layer) * 2 * parameters->batch_size *
+                                                                                   parameters->num_beams * gpt_subgraph_.num_heads *
+                                                                                   parameters->max_length * gpt_subgraph_.head_size,
+
+                                                                               2 * parameters->batch_size *
+                                                                                   parameters->num_beams * gpt_subgraph_.num_heads *
+                                                                                   parameters->sequence_length *
+                                                                                   gpt_subgraph_.head_size);
+
+      int64_t dims[] = {2, parameters->batch_size * parameters->num_beams,
```

Comment:
How does it work with Lei's past_present_share_buffer?

Reply:
It doesn't. Lei's past_present_buffer feature is only for GreedySearch/Sampling and not BeamSearch. This optimization is specific to BeamSearch where we need to select the past states of the chosen beams. 

In order to use a version of the reuse_past_present_buffer in BeamSearch, we would have to implement the Decoder Self Attention kernel which uses the cache indirection mechanism to smartly pick the past state in the Attention implementation.